sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -127,6 +127,176 @@ case class MapKeys(child: Expression)
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val storedArrTypes = ctx.freshName("storedArrTypes")
+    val returnNull = ctx.freshName("returnNull")
+    val evals = children.map(_.genCode(ctx))
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |  $returnNull[0] = true;
+        |}
+        |$storedArrTypes[$index] = "${arrayElementTypes(index)}";
+        |$biggestCardinality = Math.max($biggestCardinality, $arrCardinality[$index]);
+      """.stripMargin
+    }
+
+    val inputsSplitted = ctx.splitExpressions(
+      expressions = inputs,
+      funcName = "getInputAndCardinality",
+      returnType = "int",
+      makeSplitFunction = body =>
+        s"""
+          |$body
+          |return $biggestCardinality;
+        """.stripMargin,
+      foldFunctions = _.map(funcCall => s"$biggestCardinality = $funcCall;").mkString("\n"),
+      arguments =
+        ("ArrayData[]", arrVals) ::
+        ("int[]", arrCardinality) ::
+        ("String[]", storedArrTypes) ::
+        ("int", biggestCardinality) ::
+        ("boolean[]", returnNull) :: Nil)
+
+    val myobject = ctx.freshName("myobject")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    val cases = arrayElementTypes.distinct.map { elementType =>
+      val getArrValsItem = CodeGenerator.getValue(s"$arrVals[$j]", elementType, i)
+      s"""
+        |case "${elementType}":
+        |  $myobject[$j] = $getArrValsItem;
+        |  break;
+      """.stripMargin
+    }
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numberOfArrays];
+      |int[] $arrCardinality = new int[$numberOfArrays];
+      |int $biggestCardinality = 0;
+      |String[] $storedArrTypes = new String[$numberOfArrays];
+      |boolean[] $returnNull = new boolean[1];
+      |$returnNull[0] = false;
+      |$inputsSplitted
+      |${CodeGenerator.javaType(dataType)} ${ev.value};
+      |Object[] $args = new Object[$biggestCardinality];
+      |for (int $i = 0; $i < $biggestCardinality; $i ++) {
+      |  Object[] $myobject = new Object[$numberOfArrays];
+      |  for (int $j = 0; $j < $numberOfArrays; $j ++) {
+      |    if ($arrVals[$j] != null && $arrCardinality[$j] > $i && !$arrVals[$j].isNullAt($i)) {
+      |      switch ($storedArrTypes[$j]) {
+      |        ${cases.mkString("\n")}
+      |       default:
+      |         break;
+      |      }
+      |    } else {
+      |      $myobject[$j] = null;
+      |    }
+      |  }
+      |  $args[$i] = new $genericInternalRow($myobject);
+      |}
+      |boolean ${ev.isNull} = $returnNull[0];
+      |if (${ev.isNull}) {
+      |  ${ev.value} = null;
+      |} else {
```

Comment:
Many of the above code should be in this else-clause.
We should check there is `null`s in the inputs before the if-statement, and if there is no `null`s, create the result array.